priv/post.ex.eex

```diff
@@ -30,22 +30,21 @@ defmodule <%= context.module_name %> do
     ]
 <%= if context.protocol == "query" do %>
     input = Map.merge(input, %{"Action" => action, "Version" => "<%= context.api_version %>"})<% end %>
-    payload = <%= context.encode %>
+    payload = encode!(client, input)
     headers = AWS.Request.sign_v4(client, "POST", url, headers, payload)
+    post(client, url, payload, headers, options)
+  end
 
-    case HTTPoison.post(url, payload, headers, options) do
-      {:ok, %HTTPoison.Response{status_code: 200, body: ""} = response} ->
-        {:ok, nil, response}
-
-      {:ok, %HTTPoison.Response{status_code: 200, body: body} = response} ->
-        {:ok, <%= context.decode %>, response}
+  defp post(client, url, payload, headers, options) do
+    case AWS.Client.request(client, :post, url, payload, headers, options) do
+      {:ok, %{status_code: 200, body: body} = response} ->
+        body = if body != "", do: decode!(client, body)
+        {:ok, body, response}
 
-      {:ok, %HTTPoison.Response{body: body}} ->
-        error = <%= context.decode %>
-        {:error, error}
+      {:ok, %{body: body} = response} ->
+        {:error, decode!(client, body), response}
```

Comment:
Another benefit of {:error, {:invalid_response, whole_response}} is that we keep the same format on all error cases. We also don’t need to decode (there is even no guarantee we can decode a failure response).

Reply:
Agreed.